### PR TITLE
Empty label fields in policies are ignored

### PR DESCRIPTION
**Pull request by zyjjay**
Fix parsing issue when policy label fields are set to empty strings

ref: https://issues.redhat.com/browse/ACM-5398

### CODE REVIEW

File: controllers/configurationpolicy_utils.go [hunk 1]
```diff
@@ -198,11 +198,8 @@ func checkFieldsWithSort(mergedObj map[string]interface{}, oldObj map[string]int
 			mQty, err := apiRes.ParseQuantity(mVal)
 			if err != nil {
 				oVal := oldObj[i]
-				if oVal == nil {
```
Comment (yiraeChristineKim) [inline controllers/configurationpolicy_utils.go:201]: Why did you remove this?

File: test/e2e/case33_empty_label_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+					node, _ := clientManaged.CoreV1().Nodes().Get(context.TODO(), case33NodeName, metav1.GetOptions{})
+					labelMap := node.GetLabels()
+					found := false
+
+					_, ok := labelMap[case33StorageLabel]
+					if ok {
+						found = true
+					}
+
+					return found
```
Comment (yiraeChristineKim) [inline test/e2e/case33_empty_label_test.go:58]:
```suggestion

					_, ok := labelMap[case33StorageLabel]
					if ok {
						return ture
					}

					return false
```

**Review by yiraeChristineKim — CHANGES_REQUESTED** [review]

File: test/e2e/case33_empty_label_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+				}, defaultTimeoutSeconds, 1).Should(BeTrue())
+
+				By("verifying label field is correct")
+				Eventually(func() interface{} {
+					node, _ := clientManaged.CoreV1().Nodes().Get(context.TODO(), case33NodeName, metav1.GetOptions{})
+					labelMap := node.GetLabels()
+					merged := labelMap[case33StorageLabel]
```
Comment (yiraeChristineKim) [inline test/e2e/case33_empty_label_test.go:65]:
```suggestion
					ok, _ := labelMap[case33StorageLabel]
					return ok
```
We wanted to verify whether the label exist or not

File: kubeconfig_test-managed-control-plane [hunk 4]
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
```
Comment (yiraeChristineKim) [inline kubeconfig_test-managed-control-plane:1]: Where this file used? cannot find

File: controllers/configurationpolicy_utils.go (hunk 1 above)
Comment (zyjjay) [reply controllers/configurationpolicy_utils.go:201]: When the label that we want to add doesn't exist in the current object, `oVal` is set to `nil`. If we set `oVal` to an empty string here, the comparison `oVal != mVal` will always evaluate to `false` when `mVal` is set to an empty string as well. So the controller never emits an update event on the object. I think it's important to retain the  `nil` value as a way to detect fields that don't exist in the current object.

File: test/e2e/case33_empty_label_test.go (hunk 3 above)
Comment (zyjjay) [reply test/e2e/case33_empty_label_test.go:65]: I verified it in the `Eventually` block above this. This check is to make sure the value of the field is an empty string.

File: kubeconfig_test-managed-control-plane (hunk 4 above)
Comment (zyjjay) [reply kubeconfig_test-managed-control-plane:1]: Oh I have no idea what this is either, I will fix up my PR to remove this.

File: test/e2e/case33_empty_label_test.go (hunk 2 above)
Comment (dhaiducek) [reply test/e2e/case33_empty_label_test.go:58]: Alternatively, this would probably be better handled by the `HaveKey()` Gomega assertion rather than returning a vague boolean.

You could even use `HaveKeyWithValue()` to assert both the key and the value to combine both of this and the next test.

https://onsi.github.io/gomega/#havekeykey-interface

File: controllers/configurationpolicy_utils.go (hunk 1 above)
Comment (yiraeChristineKim) [reply controllers/configurationpolicy_utils.go:201]: @JustinKuli Would you help review this?

File: test/resources/case33_empty_label_handling/empty-labels-config.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  severity: high
+  object-templates:
+  - complianceType: musthave
+    objectDefinition:
+      apiVersion: v1
+      kind: Node
+      metadata:
+        name: test-managed-control-plane
+        labels:
+          cluster.ocs.openshift.io/openshift-storage: ""
```
Comment (dhaiducek) [inline test/resources/case33_empty_label_handling/empty-labels-config.yaml:16]: As a sanity check, could we also check using a policy that sets `cluster.ocs.openshift.io/openshift-storage: null` to make sure the policy doesn't remove the field? (Though I'm also uncertain what the behavior would be there in terms of compliance...)

File: controllers/configurationpolicy_utils.go (hunk 1 above)
Comment (JustinKuli) [reply controllers/configurationpolicy_utils.go:201]: I think I'd prefer
```
				oVal, keyPresent := oldObj[i]
				if !keyPresent {
					return false
				}
```

But I think either would work.

File: test/e2e/case33_empty_label_test.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"open-cluster-management.io/config-policy-controller/test/utils"
+)
+
+const (
+	case33ConfigPolicy     = "../resources/case33_empty_label_handling/empty-labels-config.yaml"
+	case33ConfigPolicyName = "case33-empty-labels"
+	case33NodeName         = "test-managed-control-plane"
```
Comment (JustinKuli) [inline test/e2e/case33_empty_label_test.go:19]: I'd prefer to use a different resource for the test. This relies on using a kind cluster of a specific name. If the test creates a configmap or something, the policy could just label that.

File: test/resources/case33_empty_label_handling/empty-labels-config.yaml (hunk 5 above)
Comment (JustinKuli) [reply test/resources/case33_empty_label_handling/empty-labels-config.yaml:16]: It would be nice to have a test just to check what this currently does, and understand if that behavior has changed (or if we change it accidentally in the future)

Comment (zyjjay) [reply test/resources/case33_empty_label_handling/empty-labels-config.yaml:16]: Currently, having an unset label/setting it to `nil` results in the deletion of **all** the labels attached to the object. I'll probably make a separate ticket to address this issue as it's a bit more involved and possibly affect annotations as well (need to confirm this but they both use the same helper function to populate the fields).

**Review by JustinKuli — DISMISSED** [review]
/hold

Looks good to me! Thanks for the changes to the tests, it looks great now. Just holding for any other reviewer comments.

File: test/e2e/case33_empty_label_test.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+var _ = Describe("Test resource creation when there are empty labels in configurationPolicy", Ordered, func() {
+	const (
+		case33ConfigPolicy     = "../resources/case33_empty_label_handling/empty-labels-config.yaml"
+		case33ConfigMapYaml    = "../resources/case33_empty_label_handling/configmap.yaml"
+		case33ConfigPolicyName = "case33-empty-labels"
+		case33ConfigMapName    = "case33-configmap"
+		case33Label            = "new-label"
+	)
+
+	It("verifies the configmap"+case33ConfigMapName+" exists in "+testNamespace, func() {
```
Comment (yiraeChristineKim) [inline test/e2e/case33_empty_label_test.go:25]: I don't understand why we need this step

Comment (zyjjay) [reply test/e2e/case33_empty_label_test.go:25]: The issue that the customer was facing was that when applying the label to a preexisting `Node` object, the empty labels are ignored and not applied. I'm just replicating a similar scenario except using `ConfigMaps` instead of a `Node` object.

**Review by mprahl — COMMENTED** [review]
This looks great! I just left a minor comment in the test.

File: test/e2e/case33_empty_label_test.go (hunk 7 above)
Comment (yiraeChristineKim) [reply test/e2e/case33_empty_label_test.go:25]: Oh now understand, The precondition is the existing object. Thanks @zyjjay

**Review by yiraeChristineKim — APPROVED** [review]

**Review by JustinKuli — APPROVED** [review]

**Comment by mprahl** [comment]
@zyjjay the commit title should start with a verb in the imperative mood (i.e. Fix ....). Please do this next time.

**Review by mprahl — APPROVED** [review]
/unhold

**Comment by mprahl** [comment]
> /cherry-pick release-2.7

You have to do that on the Stolostron PR:
https://github.com/stolostron/config-policy-controller/pull/568
